File: SteamLanguageParser/src/main/java/in/dragonbra/steamlanguageparser/parser/LanguageParser.java
package in.dragonbra.steamlanguageparser.parser;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * @author lngtr
 * @since 2018-02-15
 */
public class LanguageParser {
    private static final Pattern PATTERN = Pattern.compile("" +
            "(?<whitespace>\\s+)|" +
            "(?<terminator>[;])|" +
            "[\"](?<string>.+?)[\"]|" +
            "\\/\\/(?<comment>.*)$|" +
            "(?<identifier>-?[a-zA-Z_0-9][a-zA-Z0-9_:.]*)|" +
            "[#](?<preprocess>[a-zA-Z]*)|" +
            "(?<operator>[{}<>\\]=|])|" +
            "(?<invalid>[^\\s]+)", Pattern.MULTILINE    );

    private static final Collection<String> GROUP_NAMES;

    static {
        List<String> groups = new ArrayList<>();
        try {
            Method namedGroupsMethod;
            namedGroupsMethod = Pattern.class.getDeclaredMethod("namedGroups");
            namedGroupsMethod.setAccessible(true);


            Map<String, Integer> namedGroups = null;
            namedGroups = (Map<String, Integer>) namedGroupsMethod.invoke(PATTERN);


            if (namedGroups != null) {
                groups.addAll(namedGroups.keySet());

            }
        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
            e.printStackTrace();
        }

        GROUP_NAMES = Collections.unmodifiableList(groups);
    }

    public static Queue<Token> tokenizeString(String buffer) {
        return tokenizeString(buffer, "");
    }

    public static Queue<Token> tokenizeString(String buffer, String fileName) {
        String[] bufferLines = buffer.split("[\\r\\n]+");

        Queue<Token> tokens = new ArrayDeque<>();

        for (int i = 0; i < bufferLines.length; i++) {
            String line = bufferLines[i];

            Matcher matcher = PATTERN.matcher(line);

            while (matcher.find()) {
                String matchValue = null;
                String groupName = null;

                for (String tempName : GROUP_NAMES) {
                    matchValue = matcher.group(tempName);
                    groupName = tempName;
                    if (matchValue != null) {
                        break;
                    }
                }

                if (matchValue == null || "comment".equals(groupName) || "whitespace".equals(groupName)) {
                    continue;
                }

                int startColumnNumber = line.indexOf(matchValue);
                int endColumnNumber = line.indexOf(matchValue) + matchValue.length();

                TokenSourceInfo source = new TokenSourceInfo(fileName, i, startColumnNumber, i, endColumnNumber);
                Token token = new Token(groupName, matchValue, source);

                tokens.add(token);
            }
        }

        return tokens;
    }
}

File: buildSrc/src/main/kotlin/in/dragonbra/generators/steamlanguage/parser/LanguageParser.kt
package `in`.dragonbra.generators.steamlanguage.parser

import java.util.ArrayDeque
import java.util.Queue
import `in`.dragonbra.generators.steamlanguage.parser.token.TokenSourceInfo
import `in`.dragonbra.generators.steamlanguage.parser.token.Token
import java.util.regex.Pattern

class LanguageParser {
    companion object {
        private val PATTERN: Pattern = Pattern.compile(
            """(?<whitespace>\s+)|""" +
                """(?<terminator>[;])|""" +
                """["](?<string>.+?)["]|""" +
                """//(?<comment>.*)$|""" +
                """(?<identifier>-?[a-zA-Z_0-9][a-zA-Z0-9_:.]*)|""" +
                """[#](?<preprocess>[a-zA-Z]*)|""" +
                """(?<operator>[{}<>\]=|])|""" +
                """(?<invalid>[^\s]+)""",
            Pattern.MULTILINE
        )

        private val GROUP_NAMES = listOf(
            "whitespace",
            "terminator",
            "string",
            "comment",
            "identifier",
            "preprocess",
            "operator",
            "invalid"
        )

        fun tokenizeString(buffer: String, fileName: String): Queue<Token> {
            val bufferLines = buffer.split("[\\r\\n]+")
            val tokens = ArrayDeque<Token>()

            bufferLines.forEachIndexed { index, line ->
                val matcher = PATTERN.matcher(line)

                while (matcher.find()) {
                    var matchValue: String? = null
                    var groupName: String? = null

                    for (tempName in GROUP_NAMES) {
                        matchValue = matcher.group(tempName)
                        groupName = tempName
                        if (matchValue != null) {
                            break
                        }
                    }

                    if (matchValue == null || groupName == "comment" || groupName == "whitespace") {
                        continue
                    }

                    val startColumnNumber = line.indexOf(matchValue)
                    val endColumnNumber = startColumnNumber + matchValue.length
                    val source = TokenSourceInfo(fileName, index, startColumnNumber, index, endColumnNumber)
                    val token = Token(groupName, matchValue, source)

                    tokens.add(token)
                }
            }

            return tokens
        }
    }
}
